keps/0035-json-schema-export.md

```diff
@@ -0,0 +1,265 @@
+            * [Story 1](#story-1)
+         * [Implementation Details](#implementation-details)
+            * [Additional Types:](#additional-types)
+            * [Additional Attributes](#additional-attributes)
+            * [Groups](#groups)
+         * [JSON-Schema export](#json-schema-export-1)
+      * [Alternatives](#alternatives)
+      * [Implementation History](#implementation-history)
+
+<!-- Added by: aneumann, at: Tue Sep 29 15:37:14 CEST 2020 -->
+
+<!--te-->
+
+## Summary
+
+Extend the existing `params.yaml` with metadata and new types and add an export as JSON-schema.
+
+## Motivation
+
+JSON-schema is the preferred way to generate a user-friendly UI to configure a new instance of a KUDO operator or to trigger plans with parameters. KUDO should be able to export a JSON-schema that describes the required parameters.
+
+KEP-33 proposes a full replacement of the existing flat list `params.yaml` with a nested and typed structure described by JSON-schema. This is a big breaking change and requires a lot of work to get right in a single release.
+
+This KEP proposes to enhance the flat list parameters with additional attributes and add support to export a JSON-schema. This is a first step towards KEP-33, it allows tooling to be build based on JSON-Schema while not introducing breaking changes on the operator level.
+
+### Goals
+
+- Add additional attributes required for a useful JSON-schema and UI generation
+- Add support to export a JSON-schema that describes the parameters
+- Prepare Implementation of KEP-33 in the future 
+
+### Non-Goals
+
+- Handle multiple CRD versions
+- Replace the `params.yaml` with a json-schema
+- Add validation for parameters
+
+## Proposal
+
+Add new types as well as optional new attributes to enhance the parameters and bring them closer to JSON-Schema.
+
+Implement an export that generates a valid JSON-Schema out of the flat parameter list.
+
+### User Stories
+
+#### Story 1
+
+A user wants to generate a UI to install a KUDO operator. He wants to use an automatic generator, for example https://rjsf-team.github.io/react-jsonschema-form/
+
+### Implementation Details
+
+#### Additional Types: 
+
+To bring the parameters up to par with JSON-schema, we add these additional types:
+- `boolean` A "true" or "false" value, from the JSON "true" or "false" productions
+- `number` An arbitrary-precision, base-10 decimal number value, from the JSON "number" production
+
+#### Additional Attributes
+
+JSON-Schema already provides good support for generated UIs, but it can be enhanced. Additionally, we do not want to adopt the full nested structure with alternatives and references yet, there is a different KEP for that. The additional attributes should provide a step in the direction to generate a good UI from it
+
+All of the additional attributes are fully optional.
+
+`group` 
+An attribute of type "string". This attribute describes a top-level group used for this parameter.
```

Comment:
so this is just one level group?

Reply:
Yes-ish. It would probably be easy enough to allow multiple levels by separating the groups by `/` or `.`, but for now only one level is supported. Although it might be a good idea to reserve a separation character already. 
Any preferences?